server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,345 @@
+    }
+
+    /**
+     * Returns names of all files stored in this directory in sorted order
+     * Does not include locally stored block files (having _block_ in their names)
+     *
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public String[] listAll() throws IOException {
+        logger.trace("listAll() called ...");
+        readLock.lock();
+        try {
+            String[] localFiles = localDirectory.listAll();
+            logger.trace("LocalDirectory files : {}", () -> Arrays.toString(localFiles));
+            Set<String> allFiles = new HashSet<>(Arrays.asList(localFiles));
+            if (remoteDirectory != null) {
+                String[] remoteFiles = getRemoteFiles();
+                allFiles.addAll(Arrays.asList(remoteFiles));
+                logger.trace("Remote Directory files : {}", () -> Arrays.toString(remoteFiles));
+            }
+            Set<String> localLuceneFiles = allFiles.stream()
+                .filter(file -> !FileType.isBlockFile(file))
+                .collect(Collectors.toUnmodifiableSet());
+            String[] files = new String[localLuceneFiles.size()];
+            localLuceneFiles.toArray(files);
+            Arrays.sort(files);
+            logger.trace("listAll() returns : {}", () -> Arrays.toString(files));
+            return files;
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    /**
+     * Removes an existing file in the directory.
+     * Throws {@link NoSuchFileException} or {@link FileNotFoundException} in case file is not present locally and in remote as well
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void deleteFile(String name) throws IOException {
+        logger.trace("deleteFile() called {}", name);
+        writeLock.lock();
+        try {
+            localDirectory.deleteFile(name);
+            fileCache.remove(localDirectory.getDirectory().resolve(name));
+        } catch (NoSuchFileException | FileNotFoundException e) {
+            logger.trace("File {} not found in local, trying to delete from Remote", name);
+            try {
+                remoteDirectory.deleteFile(name);
+            } finally {
+                writeLock.unlock();
+            }
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Returns the byte length of a file in the directory.
+     * Throws {@link NoSuchFileException} or {@link FileNotFoundException} in case file is not present locally and in remote as well
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public long fileLength(String name) throws IOException {
+        logger.trace("fileLength() called {}", name);
+        readLock.lock();
+        try {
+            long fileLength;
+            if (Arrays.asList(getRemoteFiles()).contains(name) == false) {
```

Comment:
Would this not result in a race where-in we enter the if block and try to fetch the length from `localDirectory` but at the very moment the file gets *evicted* from `localDirectory` and is now available on the `remoteDirectory`?

Similar comment applies to other methods such as `openInput()`, `sync()`.

Reply:
Currently we are checking if the file exists in local from the acache itself. If it exists the refCount is incremented for that key and it cannot be evicted parellely. So race condition won't occur.